### DISCUSSION TITLE
CHIPS OT - Avoiding the requirement to have OT headers in every response

### DISCUSSION THREAD

**Discussion by nikhiljosya** [privacysandbox/privacy-sandbox-dev-support#33] [Phasing out third-party cookies]
### Background

My team and I have been working on using the CHIPS OT to run experiments for several features in our and our client teams that currently rely on 3P cookies. Working in a large org, we have a lot of subdomains, and most of the subdomains are served by various front-end servers. Also, the CHIPS OT uses **(a)**  the `Origin-Trial` & `Accept-CH` headers in the response to register a domains interest in participating in the OT & **(b)** expects the domains to send the `Origin-Trial` & `Accept-CH` headers in **every** subsequent response from the domain to track the intent of continued participation in the OT.

### Issue
Because of the above strict requirements w.r.t to the OT headers in every response from the registered domains, we are currently experience the following issues:
1. A complex task of identifying all the front-end servers that server requests sent to the registered domains so that we can include the headers in every response
2. Adding temporary hard-coded logic in our many front-end servers to include the OT headers in every response
3. Unstable experiments that does not provide enough valuable data to determine the effectiveness of the partitioned cookies, because our partitioned cookies might just be converted un-partitioned 3P cookies if a single response does not contain the required OT headers

### Ask
An alternative solution to the requirement of including the OT headers in every response. This would really simplify our implementation and make our experiments stable.

One possible solution could be to introduce an explicit opt-out logic for interested domains so that they do not need to send OT headers in every response but just send an opt-out header to signify the intent to opt-out.

**Comment by DCtheTall** [comment]
Hey @nikhiljosya, below is an early proposal for a new CHIPS OT design that we would support with the current design concurrently. The design is a WIP and I haven't scoped all the changes I'd need to make in Chromium, but, in theory, it should be possible.

## Which responses does the CHIPS OT really care about?

Really, CHIPS only cares about responses that have some signal that the server intends to modify the client's cookie jar. The clearest signal of this is the `Set-Cookie` header. A way to refine this even further is to only consider `Set-Cookie` headers that only contain the `Partitioned` attribute.

## New OT design

Ideally, we would want to support both the existing OT design using client hints and this new OT design simultaneously so we do not pull the rug out from partners already in the trial. Given that constraint, I propose we support two designs simultaneously:

The first is the existing OT design, so I won't go into detail about it in this doc. The design is covered in detail at [the CHIPS documentation page on chromium.org](https://chromium.org/updates/chips#origin-trial).

The second design does away with the `Accept-CH: Sec-CH-Partitioned-Cookies` header. Instead, sites would only send the `Origin-Trial` header with their CHIPS OT token when they are sending responses that include the `Set-Cookie` header with the `Partitioned` attribute. 

In the second design, Chrome would only have a signal that the site is in the OT once they have saved partitioned cookies. If the client has partitioned cookies saved, it will send the `Sec-CH-Partitioned-Cookies: ?0` request header along with the Cookie header containing the partitioned cookies. It will never send the `Sec-CH-Partitioned-Cookies: ?1` header.

If the site does not have any partitioned cookies in their cookie jar, Chrome will assume they are not in the OT, and it will not send the client hint.

### Proposed algorithms for requests/responses

Chrome will handle requests using the following algorithm (unchanged from the existing design):
1.  Check the Accept-CH cache for Sec-CH-Partitioned-Cookies, if it is in the cache, attach the `Sec-CH-Partitioned-Cookies: ?1` request header and skip the next step.
1.  Otherwise, only add `Sec-CH-Partitioned-Cookies: ?0` header if the client will attach partitioned cookies to the request.

Chrome will handle responses using the following algorithm:
1.  If the response includes `Accept-CH: Sec-CH-Partitioned` cookies, Chrome will always validate the token in the `Origin-Trial` header to verify OT participation.
1.  If Sec-CH-Partitioned-Cookies is in the Accept-CH cache but the response does not include the `Origin-Trial` **or** `Accept-CH` headers, then convert partitioned cookies to unpartitioned (this would be a site opting out of the existing design).
1.  Otherwise, check if the response has a `Set-Cookie` header with the `Partitioned` attribute. If the response does not have that header, skip the following step.
1.  If the response has a `Set-Cookie` header with the `Partitioned` attribute, then check for the presence of the `Origin-Trial` header and validate the OT token. If either the header is missing or the token is not valid, convert the site's partitioned cookies to unpartitioned (this would be the site opting out of the new design).

Let us know if this design works better for you, and open to any suggestions/feedback from you or anyone else interested in this thread.
